File: app/src/main/java/com/apps/adrcotfas/goodtime/Settings/reminders/ReminderHelper.java
package com.apps.adrcotfas.goodtime.Settings.reminders;

import android.annotation.TargetApi;
import android.app.AlarmManager;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.ComponentName;
import android.content.Context;
import android.content.ContextWrapper;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build;
import android.util.Log;

import androidx.core.app.NotificationCompat;
import androidx.preference.PreferenceManager;

import com.apps.adrcotfas.goodtime.Main.TimerActivity;
import com.apps.adrcotfas.goodtime.R;
import com.apps.adrcotfas.goodtime.Settings.PreferenceHelper;
import com.apps.adrcotfas.goodtime.Util.StringUtils;

import org.joda.time.DateTime;
import org.joda.time.LocalTime;


import static android.app.AlarmManager.RTC_WAKEUP;

import java.util.List;

public class ReminderHelper extends ContextWrapper implements SharedPreferences.OnSharedPreferenceChangeListener {

    private static final String TAG = "ReminderHelper";

    private static final String GOODTIME_REMINDER_NOTIFICATION = "goodtime_reminder_notification";
    public final static String REMINDER_ACTION = "goodtime.reminder_action";
    public final static int REMINDER_REQUEST_CODE = 11;
    public static final int REMINDER_NOTIFICATION_ID = 99;

    private static AlarmManager alarmManager;
    private static PendingIntent pendingIntent;

    public ReminderHelper(Context context) {
        super(context);
        PreferenceManager.getDefaultSharedPreferences(context).registerOnSharedPreferenceChangeListener(this);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            initChannel();
        }
        if (PreferenceHelper.isReminderEnabled()) {
            enableBootReceiver();
            scheduleNotification(context);
        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    private void initChannel() {
        Log.d(TAG, "initChannel");
        NotificationChannel c = new NotificationChannel(
                GOODTIME_REMINDER_NOTIFICATION,
                getString(R.string.reminder_notification_channel_name),
                NotificationManager.IMPORTANCE_DEFAULT);
        c.setShowBadge(true);
        NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        notificationManager.createNotificationChannel(c);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (key.equals(PreferenceHelper.ENABLE_REMINDER)) {
            Log.d(TAG, "onSharedPreferenceChanged");
            if (PreferenceHelper.isReminderEnabled()) {
                enableBootReceiver();
                scheduleNotification(getApplicationContext());
            } else {
                disableBootReceiver();
                unscheduledNotification(getApplicationContext());
            }
        }
    }

    public void enableBootReceiver() {
        Log.d(TAG, "enableBootReceiver");
        ComponentName receiver = new ComponentName(this, BootReceiver.class);
        PackageManager pm = this.getPackageManager();
        pm.setComponentEnabledSetting(receiver,
                PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
                PackageManager.DONT_KILL_APP);
    }

    private void disableBootReceiver() {
        Log.d(TAG, "disableBootReceiver");
        ComponentName receiver = new ComponentName(this, BootReceiver.class);
        PackageManager pm = this.getPackageManager();
        pm.setComponentEnabledSetting(receiver,
                PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
                PackageManager.DONT_KILL_APP);
    }

    private AlarmManager getAlarmManager() {
        if (alarmManager  == null) {
            alarmManager = (AlarmManager) this.getSystemService(ALARM_SERVICE);
        }
        return alarmManager;
    }

    private static PendingIntent getReminderPendingIntent(Context context) {
        if (pendingIntent == null) {
            Intent intent = new Intent(context, ReminderReceiver.class);
            intent.setAction(REMINDER_ACTION);
            pendingIntent = PendingIntent.getBroadcast(
                    context,
                    REMINDER_REQUEST_CODE,
                    intent,
                    PendingIntent.FLAG_IMMUTABLE);
        }
        return pendingIntent;
    }

    private void unscheduledNotification(Context context) {
        Log.d(TAG, "unscheduledNotification");
        getAlarmManager().cancel(getReminderPendingIntent(context));
    }

    public static void scheduleNotification(Context context) {
        if (PreferenceHelper.isReminderEnabled()) {

            long currentUpcomingReminderMillis = PreferenceHelper.getTimeOfReminder();
            List<Integer> weekdaysOfReminder = PreferenceHelper.getWeekdaysOfReminder();
            Log.d(TAG, "time of reminder: " + StringUtils.formatDateAndTime(currentUpcomingReminderMillis));

            DateTime reminderTime = new DateTime(currentUpcomingReminderMillis);

            long nextReminderMillis = UpcomingReminder.Companion.calculateNextAlertInMillis(
                    reminderTime.getHourOfDay(),
                    reminderTime.getMinuteOfHour(),
                    weekdaysOfReminder
            );

            Log.d(TAG, "scheduleNotification at: " + StringUtils.formatDateAndTime(nextReminderMillis));

            AlarmManager alarmManager = (AlarmManager) context.getSystemService(ALARM_SERVICE);
            alarmManager.set(
                    RTC_WAKEUP,
                    nextReminderMillis,
                    getReminderPendingIntent(context));
        }
    }

    public static void notifyReminder(Context context) {
        Intent openMainIntent = new Intent(context, TimerActivity.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, openMainIntent, PendingIntent.FLAG_IMMUTABLE);

        final NotificationCompat.Builder builder =
                new NotificationCompat.Builder(context, GOODTIME_REMINDER_NOTIFICATION)
                        .setSmallIcon(R.drawable.ic_status_goodtime)
                        .setCategory(NotificationCompat.CATEGORY_REMINDER)
                        .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                        .setContentIntent(pendingIntent)
                        .setShowWhen(false)
                        .setOnlyAlertOnce(true)
                        .setContentTitle(context.getString(R.string.reminder_title))
                        .setContentText(context.getString(R.string.reminder_text));
        final NotificationManager notificationManager =
                (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        notificationManager.notify(REMINDER_NOTIFICATION_ID, builder.build());
    }

    public static void removeNotification(Context context) {
        final NotificationManager notificationManager =
                (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        notificationManager.cancel(REMINDER_NOTIFICATION_ID);
    }
}

File: app/src/main/java/com/apps/adrcotfas/goodtime/settings/reminders/ReminderHelper.kt
package com.apps.adrcotfas.goodtime.settings.reminders

import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.annotation.TargetApi
import android.os.Build
import android.app.NotificationChannel
import com.apps.adrcotfas.goodtime.R
import android.app.NotificationManager
import com.apps.adrcotfas.goodtime.settings.PreferenceHelper
import android.content.pm.PackageManager
import android.app.AlarmManager
import android.app.PendingIntent
import android.content.*
import android.content.Context.ALARM_SERVICE
import android.content.Context.NOTIFICATION_SERVICE
import android.util.Log
import androidx.core.app.NotificationCompat
import com.apps.adrcotfas.goodtime.main.TimerActivity
import com.apps.adrcotfas.goodtime.util.StringUtils.formatDateAndTime
import dagger.hilt.android.qualifiers.ApplicationContext
import java.time.*
import java.time.temporal.TemporalAdjusters
import javax.inject.Inject

class ReminderHelper@Inject constructor(
    @ApplicationContext val context: Context,
    val preferenceHelper: PreferenceHelper
) : OnSharedPreferenceChangeListener {

    private var pendingIntents : Array<PendingIntent?> = arrayOfNulls(7)
    private val alarmManager: AlarmManager by lazy {
        context.getSystemService(ALARM_SERVICE) as AlarmManager
    }

    init {
        preferenceHelper.preferences.registerOnSharedPreferenceChangeListener(this)
        initChannelIfNeeded()
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun initChannelIfNeeded() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val notificationManager = context.getSystemService(NOTIFICATION_SERVICE) as NotificationManager
            val channel = notificationManager.getNotificationChannel(GOODTIME_REMINDER_CHANNEL_ID)
            if (channel == null) {
                Log.d(TAG, "initChannel")
                val c = NotificationChannel(
                    GOODTIME_REMINDER_CHANNEL_ID,
                    context.getString(R.string.reminder_notification_channel_name),
                    NotificationManager.IMPORTANCE_DEFAULT
                )
                c.setShowBadge(true)
                notificationManager.createNotificationChannel(c)
            }
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String?) {
        // some weird crash in the pre-report required this
        if (key == null) {
            return
        }
        if (key.contains(PreferenceHelper.REMINDER_DAYS)) {
            Log.d(TAG, "onSharedPreferenceChanged: $key")
            val idx = key.last().toInt() - '0'.toInt()
            val reminderDay = DayOfWeek.of(idx + 1)
            if (preferenceHelper.isReminderEnabledFor(reminderDay)) {
                toggleBootReceiver(true)
                scheduleNotification(reminderDay)
            } else {
                cancelNotification(reminderDay)
                toggleBootReceiver(false)
            }
        } else if (key == PreferenceHelper.REMINDER_TIME) {
            Log.d(TAG, "onSharedPreferenceChanged: REMINDER_TIME")
            cancelNotifications()
            scheduleNotifications()
        }
    }

    private fun toggleBootReceiver(enabled: Boolean) {
        Log.d(TAG, "toggleBootReceiver ${if (enabled) "ENABLED" else "DISABLED"}")
        val receiver = ComponentName(context, BootReceiver::class.java)
        val pm = context.packageManager
        pm.setComponentEnabledSetting(
            receiver,
            if (enabled) PackageManager.COMPONENT_ENABLED_STATE_ENABLED
            else PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
            PackageManager.DONT_KILL_APP
        )
    }

    private fun getReminderPendingIntent(index: Int): PendingIntent {
        if (pendingIntents[index] == null) {
            val intent = Intent(context, ReminderReceiver::class.java)
            intent.action = REMINDER_ACTION
            pendingIntents[index] = PendingIntent.getBroadcast(
                context,
                REMINDER_REQUEST_CODE + index,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )
        }
        return pendingIntents[index]!!
    }

    private fun cancelNotifications() {
        Log.d(TAG, "cancelNotifications")
        for (day in DayOfWeek.values()) {
            cancelNotification(day)
        }
    }

    private fun cancelNotification(day: DayOfWeek) {
        Log.d(TAG, "cancelNotification for $day")
        val reminderPendingIntent = getReminderPendingIntent(day.ordinal)
        alarmManager.cancel(reminderPendingIntent)
    }

    fun scheduleNotifications() {
        if (preferenceHelper.isReminderEnabled()) {
            for (i in preferenceHelper.getReminderDays().withIndex()) {
                if (i.value) {
                    val reminderDay = DayOfWeek.of(i.index + 1)
                    scheduleNotification(reminderDay)
                }
            }
        }
    }

    private fun scheduleNotification(reminderDay: DayOfWeek) {
        val now = LocalDateTime.now()
        Log.d(TAG, "now: ${now.toLocalTime()}")

        val time = LocalTime.ofSecondOfDay(preferenceHelper.getReminderTime().toLong())
        Log.d(TAG, "time of reminder: $time")

        var reminderTime = now
            .withHour(time.hour)
            .withMinute(time.minute)
            .withSecond(0)
            .with(TemporalAdjusters.nextOrSame(reminderDay))

        if (reminderTime.isBefore(now)) {
            Log.d(TAG, "reminderTime is before now; schedule for next week")
            reminderTime = reminderTime.plusWeeks(1)
        }

        Log.d(TAG, "reminderTime: $reminderTime")

        val reminderMillis = reminderTime.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli()

        Log.d(TAG, "scheduleNotification at: " + formatDateAndTime(reminderMillis))
        alarmManager.setInexactRepeating(
            AlarmManager.RTC_WAKEUP,
            reminderMillis,
            AlarmManager.INTERVAL_DAY * 7,
            getReminderPendingIntent(reminderDay.ordinal)
        )
    }

    companion object {
        private const val TAG = "ReminderHelper"
        private const val GOODTIME_REMINDER_CHANNEL_ID = "goodtime_reminder_notification"
        const val REMINDER_ACTION = "goodtime.reminder_action"
        const val REMINDER_REQUEST_CODE = 11
        private const val REMINDER_NOTIFICATION_ID = 99

        fun notifyReminder(context: Context) {
            val openMainIntent = Intent(context, TimerActivity::class.java)
            val pendingIntent = PendingIntent.getActivity(
                context,
                0,
                openMainIntent,
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )
            val builder = NotificationCompat.Builder(context, GOODTIME_REMINDER_CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_status_goodtime)
                .setCategory(NotificationCompat.CATEGORY_REMINDER)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setContentIntent(pendingIntent)
                .setShowWhen(false)
                .setOnlyAlertOnce(true)
                .setContentTitle(context.getString(R.string.reminder_title))
                .setContentText(context.getString(R.string.reminder_text))
            val notificationManager =
                context.getSystemService(NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.notify(REMINDER_NOTIFICATION_ID, builder.build())
        }
        fun removeNotification(context: Context) {
            val notificationManager =
                context.getSystemService(NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.cancel(REMINDER_NOTIFICATION_ID)
        }
    }
}